File: src/shell/states/If.zig
pub const If = @This();

base: State,
node: *const ast.If,
parent: ParentPtr,
io: IO,
state: union(enum) {
    idle,
    exec: struct {
        state: union(enum) {
            cond,
            then,
            elif: struct {
                idx: u32 = 0,
            },
            @"else",
        },
        stmts: *const SmolList(ast.Stmt, 1),
        stmt_idx: u32 = 0,
        last_exit_code: ExitCode = 0,
        exit_requested: bool = false,
    },
    waiting_write_err,
    done,
} = .idle,

pub const ParentPtr = StatePtrUnion(.{
    Stmt,
    Binary,
    Pipeline,
    Async,
});

pub const ChildPtr = StatePtrUnion(.{
    Stmt,
});

pub fn format(this: *const If, comptime _: []const u8, _: std.fmt.FormatOptions, writer: anytype) !void {
    try writer.print("If(0x{x}, state={s})", .{ @intFromPtr(this), @tagName(this.state) });
}

pub fn init(
    interpreter: *Interpreter,
    shell_state: *ShellExecEnv,
    node: *const ast.If,
    parent: ParentPtr,
    io: IO,
) *If {
    const if_stmt = parent.create(If);
    if_stmt.* = .{
        .base = State.initWithNewAllocScope(.if_clause, interpreter, shell_state),
        .node = node,
        .parent = parent,
        .io = io,
    };
    return if_stmt;
}

pub fn start(this: *If) Yield {
    return .{ .@"if" = this };
}

pub fn next(this: *If) Yield {
    while (this.state != .done) {
        switch (this.state) {
            .idle => {
                this.state = .{ .exec = .{ .state = .cond, .stmts = &this.node.cond } };
            },
            .exec => {
                const stmts = this.state.exec.stmts;
                // Executed all the stmts in the condition/branch
                if (this.state.exec.stmt_idx >= stmts.len()) {
                    switch (this.state.exec.state) {
                        // Move to the then, elif, or else branch based on the exit code
                        // and the amount of else parts
                        .cond => {
                            if (this.state.exec.last_exit_code == 0) {
                                this.state.exec.state = .then;
                                this.state.exec.stmt_idx = 0;
                                this.state.exec.stmts = &this.node.then;
                                continue;
                            }
                            switch (this.node.else_parts.len()) {
                                0 => {
                                    return this.parent.childDone(this, 0);
                                },
                                1 => {
                                    this.state.exec.state = .@"else";
                                    this.state.exec.stmt_idx = 0;
                                    this.state.exec.stmts = this.node.else_parts.getConst(0);
                                    continue;
                                },
                                else => {
                                    this.state.exec.state = .{ .elif = .{} };
                                    this.state.exec.stmt_idx = 0;
                                    this.state.exec.stmts = this.node.else_parts.getConst(0);
                                    continue;
                                },
                            }
                        },
                        // done
                        .then => {
                            return this.parent.childDone(this, this.state.exec.last_exit_code);
                        },
                        // if succesful, execute the elif's then branch
                        // otherwise, move to the next elif, or to the final else if it exists
                        .elif => {
                            if (this.state.exec.last_exit_code == 0) {
                                this.state.exec.stmts = this.node.else_parts.getConst(this.state.exec.state.elif.idx + 1);
                                this.state.exec.stmt_idx = 0;
                                this.state.exec.state = .then;
                                continue;
                            }

                            this.state.exec.state.elif.idx += 2;

                            if (this.state.exec.state.elif.idx >= this.node.else_parts.len()) {
                                return this.parent.childDone(this, 0);
                            }

                            if (this.state.exec.state.elif.idx == this.node.else_parts.len() -| 1) {
                                this.state.exec.state = .@"else";
                                this.state.exec.stmt_idx = 0;
                                this.state.exec.stmts = this.node.else_parts.lastUncheckedConst();
                                continue;
                            }

                            this.state.exec.stmt_idx = 0;
                            this.state.exec.stmts = this.node.else_parts.getConst(this.state.exec.state.elif.idx);
                            continue;
                        },
                        .@"else" => {
                            return this.parent.childDone(this, this.state.exec.last_exit_code);
                        },
                    }
                }

                const idx = this.state.exec.stmt_idx;
                this.state.exec.stmt_idx += 1;
                const stmt = this.state.exec.stmts.getConst(idx);
                var newstmt = Stmt.init(this.base.interpreter, this.base.shell, stmt, this, this.io.copy());
                return newstmt.start();
            },
            .waiting_write_err => return .suspended, // yield execution
            .done => @panic("This code should not be reachable"),
        }
    }

    return this.parent.childDone(this, 0);
}

pub fn deinit(this: *If) void {
    log("{} deinit", .{this});
    this.io.deref();
    this.base.endScope();
    this.parent.destroy(this);
}

pub fn childDone(this: *If, child: ChildPtr, exit_code: ExitCode) Yield {
    return this.childDoneWithFlag(child, exit_code, false);
}

pub fn childDoneWithExit(this: *If, child: ChildPtr, exit_code: ExitCode) Yield {
    return this.childDoneWithFlag(child, exit_code, true);
}

fn childDoneWithFlag(this: *If, child: ChildPtr, exit_code: ExitCode, exit_requested: bool) Yield {
    // Check if child had exit
    const child_had_exit = exit_requested or brk: {
        if (child.ptr.is(Stmt)) {
            const stmt = child.as(Stmt);
            if (stmt.exit_requested) {
                break :brk true;
            }
        }
        break :brk false;
    };

    defer child.deinit();

    if (this.state != .exec) {
        @panic("Expected `exec` state in If, this indicates a bug in Bun. Please file a GitHub issue.");
    }

    var exec = &this.state.exec;
    exec.last_exit_code = exit_code;

    // If exit was requested, propagate it
    if (child_had_exit) {
        exec.exit_requested = true;
        // Propagate exit to parent
        if (this.parent.ptr.is(Stmt)) {
            return this.parent.as(Stmt).childDoneWithExit(Stmt.ChildPtr.init(this), exit_code);
        } else if (this.parent.ptr.is(Binary)) {
            return this.parent.as(Binary).childDoneWithExit(Binary.ChildPtr.init(this), exit_code);
        } else if (this.parent.ptr.is(Pipeline)) {
            // Pipeline doesn't have childDoneWithExit, just use regular childDone
            return this.parent.as(Pipeline).childDone(Pipeline.ChildPtr.init(this), exit_code);
        } else if (this.parent.ptr.is(Async)) {
            // Async doesn't have childDoneWithExit, just use regular childDone
            return this.parent.as(Async).childDone(Async.ChildPtr.init(this), exit_code);
        }
        @panic("Unexpected parent type for If");
    }

    switch (exec.state) {
        .cond => return .{ .@"if" = this },
        .then => return .{ .@"if" = this },
        .elif => {
            // if (exit_code == 0) {
            //     exec.stmts = this.node.else_parts.getConst(exec.state.elif.idx + 1);
            //     exec.state = .then;
            //     exec.stmt_idx = 0;
            //     this.next();
            //     return;
            // }
            return .{ .@"if" = this };
        },
        .@"else" => return .{ .@"if" = this },
    }
}

const bun = @import("bun");
const std = @import("std");

const shell = bun.shell;
const ExitCode = bun.shell.ExitCode;
const SmolList = bun.shell.SmolList;
const Yield = bun.shell.Yield;
const ast = bun.shell.AST;

const Interpreter = bun.shell.Interpreter;
const Async = bun.shell.Interpreter.Async;
const Binary = bun.shell.Interpreter.Binary;
const IO = bun.shell.Interpreter.IO;
const Pipeline = bun.shell.Interpreter.Pipeline;
const ShellExecEnv = Interpreter.ShellExecEnv;
const State = bun.shell.Interpreter.State;
const Stmt = bun.shell.Interpreter.Stmt;

const StatePtrUnion = bun.shell.interpret.StatePtrUnion;
const log = bun.shell.interpret.log;

File: src/runtime/shell/states/If.rs
use crate::shell::ExitCode;
use crate::shell::ast;
use crate::shell::interpreter::{Interpreter, Node, NodeId, ShellExecEnv, StateKind, log};
use crate::shell::io::IO;
use crate::shell::states::base::Base;
use crate::shell::states::stmt::Stmt;
use crate::shell::yield_::Yield;

pub struct If {
    pub base: Base,
    pub node: bun_ptr::BackRef<ast::If>,
    pub io: IO,
    pub state: IfState,
}

#[derive(Default, strum::IntoStaticStr)]
pub enum IfState {
    #[default]
    Idle,
    Exec(Exec),
    WaitingWriteErr,
    Done,
}

pub struct Exec {
    pub state: ExecBranch,
    /// Back-reference to the current `SmolList<ast::Stmt, 1>` being walked.
    /// Points into the AST arena, which the interpreter holds for its entire
    /// lifetime — it outlives every state node.
    pub stmts: bun_ptr::BackRef<ast::SmolList<ast::Stmt, 1>>,
    pub stmt_idx: u32,
    pub last_exit_code: ExitCode,
}

impl Exec {
    /// Borrow the current `SmolList<Stmt, 1>` being walked.
    ///
    /// `stmts` always points into the AST arena (`ShellArgs::__arena`), which
    /// the interpreter holds for its entire lifetime — it outlives every state
    /// node (BackRef invariant).
    #[inline]
    fn stmts(&self) -> &ast::SmolList<ast::Stmt, 1> {
        self.stmts.get()
    }

    #[inline]
    fn stmts_len(&self) -> u32 {
        self.stmts().len() as u32
    }
}

pub enum ExecBranch {
    Cond,
    Then,
    Elif { idx: u32 },
    Else,
}

impl If {
    pub fn init(
        interp: &Interpreter,
        shell: *mut ShellExecEnv,
        node: &ast::If,
        parent: NodeId,
        io: IO,
    ) -> NodeId {
        interp.alloc_node(Node::If(If {
            base: Base::new(StateKind::IfClause, parent, shell),
            node: bun_ptr::BackRef::new(node),
            io,
            state: IfState::Idle,
        }))
    }

    pub fn start(_interp: &Interpreter, this: NodeId) -> Yield {
        Yield::Next(this)
    }

    pub fn next(interp: &Interpreter, this: NodeId) -> Yield {
        let parent = interp.as_if(this).base.parent;
        loop {
            // PORT NOTE: reshaped for borrowck — we read/mutate `state` via a
            // short-lived borrow, decide an action, then drop the borrow
            // before calling back into `interp`.
            let action = {
                let me = interp.as_if_mut(this);
                // Copy the BackRef out so `n` borrows a local, leaving `me`
                // free for the disjoint `&mut me.state` borrow below.
                let node = me.node;
                let n = node.get();
                match &mut me.state {
                    IfState::Idle => {
                        me.state = IfState::Exec(Exec {
                            state: ExecBranch::Cond,
                            stmts: bun_ptr::BackRef::new(&n.cond),
                            stmt_idx: 0,
                            last_exit_code: 0,
                        });
                        continue;
                    }
                    IfState::Exec(exec) => {
                        if exec.stmt_idx >= exec.stmts_len() {
                            match &mut exec.state {
                                ExecBranch::Cond => {
                                    if exec.last_exit_code == 0 {
                                        exec.state = ExecBranch::Then;
                                        exec.stmts = bun_ptr::BackRef::new(&n.then);
                                        exec.stmt_idx = 0;
                                        continue;
                                    }
                                    let else_len = n.else_parts.len() as u32;
                                    match else_len {
                                        0 => Action::Done(0),
                                        1 => {
                                            exec.state = ExecBranch::Else;
                                            exec.stmts = bun_ptr::BackRef::new(&n.else_parts[0]);
                                            exec.stmt_idx = 0;
                                            continue;
                                        }
                                        _ => {
                                            exec.state = ExecBranch::Elif { idx: 0 };
                                            exec.stmts = bun_ptr::BackRef::new(&n.else_parts[0]);
                                            exec.stmt_idx = 0;
                                            continue;
                                        }
                                    }
                                }
                                ExecBranch::Then => Action::Done(exec.last_exit_code),
                                ExecBranch::Elif { idx } => {
                                    if exec.last_exit_code == 0 {
                                        // The matching `then` arm follows the
                                        // `elif` cond at `idx + 1`.
                                        let then_idx = *idx + 1;
                                        exec.state = ExecBranch::Then;
                                        exec.stmts =
                                            bun_ptr::BackRef::new(&n.else_parts[then_idx as usize]);
                                        exec.stmt_idx = 0;
                                        continue;
                                    }
                                    *idx += 2;
                                    let else_len = n.else_parts.len() as u32;
                                    if *idx >= else_len {
                                        Action::Done(0)
                                    } else if *idx == else_len - 1 {
                                        exec.state = ExecBranch::Else;
                                        exec.stmts = bun_ptr::BackRef::new(
                                            &n.else_parts[(else_len - 1) as usize],
                                        );
                                        exec.stmt_idx = 0;
                                        continue;
                                    } else {
                                        exec.stmts =
                                            bun_ptr::BackRef::new(&n.else_parts[*idx as usize]);
                                        exec.stmt_idx = 0;
                                        continue;
                                    }
                                }
                                ExecBranch::Else => Action::Done(exec.last_exit_code),
                            }
                        } else {
                            let i = exec.stmt_idx;
                            exec.stmt_idx += 1;
                            // `i` was bounds-checked against `stmts_len()`.
                            let stmt_node: *const ast::Stmt = &raw const exec.stmts()[i as usize];
                            Action::SpawnStmt(stmt_node)
                        }
                    }
                    IfState::WaitingWriteErr => return Yield::suspended(),
                    IfState::Done => panic!("This code should not be reachable"),
                }
            };
            return match action {
                Action::Done(exit) => interp.child_done(parent, this, exit),
                Action::SpawnStmt(stmt_node) => {
                    let (shell, io) = {
                        let me = interp.as_if(this);
                        (me.base.shell, me.io.clone())
                    };
                    let new_stmt = Stmt::init(interp, shell, stmt_node, this, io);
                    Stmt::start(interp, new_stmt)
                }
            };
        }
    }

    pub fn child_done(
        interp: &Interpreter,
        this: NodeId,
        child: NodeId,
        exit_code: ExitCode,
    ) -> Yield {
        interp.deinit_node(child);
        let me = interp.as_if_mut(this);
        let IfState::Exec(exec) = &mut me.state else {
            panic!(
                "Expected `exec` state in If, this indicates a bug in Bun. Please file a GitHub issue."
            );
        };
        exec.last_exit_code = exit_code;
        Yield::Next(this)
    }

    pub fn deinit(interp: &Interpreter, this: NodeId) {
        log!("If {} deinit", this);
        interp.as_if_mut(this).base.end_scope();
    }
}

enum Action {
    Done(ExitCode),
    SpawnStmt(*const ast::Stmt),
}

// ported from: src/shell/states/If.zig
